zio/ndjsonio/typeparser.go

```diff
@@ -0,0 +1,322 @@
+// nested record fields). The two descriptors here represent the same data
+// in the same order, flatDescriptor describes the data as it appears in
+// JSON, descriptor describes it as it appears in zng values.
+type typeInfo struct {
+	descriptor *zng.TypeRecord
+	flatDesc   *zng.TypeRecord
+	path       []byte
+	jsonVals   []jsonVal
+}
+type jsonVal struct {
+	val []byte
+	typ jsonparser.ValueType
+}
+
+func getUnsafeDefault(data []byte, defaultValue string, key string) (string, error) {
+	val, err := jsonparser.GetUnsafeString(data, key)
+	if err != nil {
+		// This is always a KeyPathNotFoundError, including if the json was invalid.
+		if defaultValue == "" {
+			return "", jsonparser.KeyPathNotFoundError
+		}
+		return defaultValue, nil
+	}
+	return val, nil
+}
+
+func lookupTypeInfo(zctx *resolver.Context, desc *zng.TypeRecord, path string) *typeInfo {
+	flatCols := zeekio.FlattenColumns(desc.Columns)
+	flatDesc := zctx.LookupTypeRecord(flatCols)
+	info := typeInfo{desc, flatDesc, []byte(path), make([]jsonVal, len(flatDesc.Columns))}
+	return &info
+}
+
+// newRawFromJSON builds a raw value from a descriptor and the JSON object
+// in data.  It works in two steps.  First, it constructs a slice of views onto
+// the underlying JSON values.  This slice follows the order of the flattened
+// columns.  Second, it builds the full encoded value and building nested
+// records as necessary.
+func (info *typeInfo) newRawFromJSON(data []byte) (zcode.Bytes, int, error) {
+	var droppedFields int
+
+	for i := range info.jsonVals {
+		info.jsonVals[i].typ = jsonparser.NotExist
+	}
+
+	// path is always the first field (typings config is validated
+	// for this, and inferred TDs are sorted with _path first).
+	info.jsonVals[0] = jsonVal{info.path, jsonparser.String}
+
+	var prefix []string
+
+	// callback can't be declared in one line due to golang/go#226
+	var callback func(key []byte, val []byte, typ jsonparser.ValueType, offset int) error
+	callback = func(key []byte, val []byte, typ jsonparser.ValueType, offset int) error {
+		skey := string(key)
+		if typ == jsonparser.Object {
+			prefix = append(prefix, skey)
+			err := jsonparser.ObjectEach(val, callback)
+			prefix = prefix[0 : len(prefix)-1]
+			return err
+		}
+
+		fullkey := strings.Join(append(prefix, skey), ".")
+
+		if col, ok := info.flatDesc.ColumnOfField(fullkey); ok {
+			info.jsonVals[col] = jsonVal{val, typ}
+		} else {
+			droppedFields++
+		}
+		return nil
+	}
+	if err := jsonparser.ObjectEach(data, callback); err != nil {
+		return nil, 0, err
+	}
+
+	builder := zcode.NewBuilder()
+	colno := 0
+	nestedColno := 0
+	nestedColumns := info.descriptor.Columns
+	flatColumns := info.flatDesc.Columns
+	for i := range flatColumns {
+		val := info.jsonVals[i].val
+		if i == info.descriptor.TsCol {
+			if info.jsonVals[i].typ != jsonparser.String && info.jsonVals[i].typ != jsonparser.Number {
+				return nil, 0, fmt.Errorf("invalid json type for ts: %s", info.jsonVals[i].typ)
+			}
```

Comment:
nit: seems like this should be inside the function parseJSONTimestamp

Reply:
meaning it would take a `jsonVal` rather than a `[]byte`? I'm not sure, I like the idea of parseJSONTimestamp being agnostic to the json typing stuff (even if it's only used here right now)

Reply:
Feels to me like it splits the logic of "what is a valid representation of a timestamp in json" between here and parseJSONTimestamp but if you prefer it this way, I don't feel strongly.